Add dtype attribute to CSEVariable

arui-meta opened the pull request:
Summary:
- This diff introduces `dtype` attribute to `TritonCSEVariable` and a dtype propagation helper function to infer dtype from input to output for each op.

- There will be a follow-up diff that uses this `dtype` information in `TritonCSEVariable` to perform dtype-aware codegen.

Test Plan: CI

Differential Revision: D61815079


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

eellison reviewed (commented):
Left a few comments - I think there is one general discussion that I tagged @blaine-rister on we should discuss.

To correctly propagate we'll need to track `is_scalar` as well - i don't think that would be too tricky, it's basically just the `constant` that should be a scalar i think. 

So long as we're not actually using the dtypes yet in inductor codegen we could potentially land this in steps but I do there's a bit more work before we can rely on this.

arui-meta commented on torch/_inductor/codegen/common.py:
```diff
@@ -1606,13 +1613,185 @@ def generate(
+        return torch.float32
+
+    @staticmethod
+    def masked(mask, body, other):
+        return torch.float32
```
Thanks! Will do!

blaine-rister replied:
Might be good to leave a TODO in the code so we don't forget :)

arui-meta commented on torch/_inductor/codegen/triton.py:
```diff
@@ -839,6 +840,8 @@ def expm1(x):
 
     @staticmethod
     def sqrt(x):
+        if x.dtype in [torch.float16, torch.bfloat16]:
+            return f"libdevice.sqrt({x}.to(tl.float32))"
         return f"libdevice.sqrt({x})"
```
Thanks for the feedback! 

It seems reasonable to always preserve the same dtype as torch by inserting upcast on input and downcast on output. We could also easily remove these casts when a given triton operator starts to support fp16/bf16 in the future.

arui-meta commented:
In the latest version, I have made the following changes since last review

- Replace try-catch block with a default dtype propagation handler
- Updated dtype promotion logic to include scalars and use `torch._prims_common.elementwise_dtypes`
- Updated `torch._prims_common.elementwise_dtypes` to work with CSEVariable
- Updated dtype_aware_codegen test to test with both float16 and bfloat16 inputs
- Cast back to input dtype when triton function call automatically upcast to float32. 

I still kept the `promote_types` calls for unary operators because sometimes we receive a str on the input argument in template codegen. Will need a separate fix before we can rely on `arg.dtype`. 
@eellison @blaine-rister

blaine-rister commented on test/inductor/test_torchinductor.py:
```diff
@@ -12145,18 +12145,35 @@ def f(x, mask):
+        @parametrize("load_upcast_to_fp32", [False, True])
+        @parametrize("input_dtype", [torch.float16, torch.bfloat16])
+        @config.patch("triton.use_block_ptr", True)
+        def test_dtype_aware_codegen(self, load_upcast_to_fp32, input_dtype):
```
Nice test. Does this PR already affect the codegen? From the description, it sounded like that would be a follow up.

arui-meta replied:
We still keep the inductor flag `triton.codegen_upcast_to_fp32 = True` by default. Codegen will still see all CSEVariables in FP32 by default. I only updated the `sqrt` codegen so that I can put in an unit test to validate dtype-aware codegen works by switching off `triton.codegen_upcast_to_fp32` and feed in lower precision inputs.

blaine-rister reviewed (commented):
LGTM! I'll have to defer to @eellison  for the final approval. The basic logic makes sense to me. I didn't verify all the different dtype propagation rules for reductions, but it sounds like those will be tested in follow-up PRs. I left comments in a few places where we might be able to streamline the codegen, once we scale this up to cover more ops.

arui-meta commented on torch/_prims_common/__init__.py:
```diff
@@ -1495,11 +1505,15 @@ def _find_highest_dtype_filtered(
                 if float_as_complex and is_float_dtype(_dtype):
                     _dtype = corresponding_complex_dtype(_dtype)
-                if x.ndim == 0:
+                if hasattr(x, "ndim") and x.ndim == 0:
```
This util function has been updated to deal with CSEVariables, which do not have an attribute "ndim"

arui-meta commented on torch/_inductor/codegen/common.py:
```diff
@@ -1747,13 +1754,198 @@ def generate(
+def promote_types(args):
+    dtype_prop_candidates = []
+
+    # CSEVariable and scalar will be included in dtype_prop_candidates
```
Regarding scalar dtypes, the pytorch type promotion util function used here is`torch._prims_common.elementwise_dtypes` , which already handles scalars in input args. Depending on the scalar dtype relative to tensor dtype, this util function will decide whether or not scalar inputs will participate in dtype propagation. 

Regarding the triton util function, there is still an upcast to float32 in the promotion logic if the operator is div or mod. https://github.com/triton-lang/triton/blob/fa229d1c4bee16c094be9427334575ec1e79f66c/python/triton/language/semantic.py#L86

blaine-rister replied:
Got it! Thanks for clarifying. LGTM.

arui-meta commented on torch/_inductor/codegen/triton.py:
```diff
@@ -849,6 +850,10 @@ def expm1(x):
 
     @staticmethod
     def sqrt(x):
+        if x.dtype == torch.float16:
```
Great idea!

eellison commented on torch/_prims_common/__init__.py:
```diff
@@ -1474,8 +1474,18 @@ def elementwise_dtypes(
     # See https://dev-discuss.pytorch.org/t/delving-into-what-happens-when-you-import-torch/1589
     import sympy
 
+    from torch._inductor.codegen.common import CSEVariable
+
```
We should not leak concepts from torchinductor into prims dtype promotion. You can avoid this by doing what inductor does [here]( https://github.com/pytorch/pytorch/blob/c272526ea54b8063c596addb48cd3a87ab91a5ef/torch/_inductor/lowering.py#L224).

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -849,7 +850,11 @@ def expm1(x):
     def sqrt(x):
-        return f"libdevice.sqrt({x})"
+        needs_upcast = x.dtype in (torch.float16, torch.bfloat16)
+        orig_dtype = str(x.dtype).replace("torch", "tl")
```
There is a method for this elsewhere in this file, please reuse it. `triton_type`

arui-meta replied:
Right now it seems that `triton_type` will map float16 / bfloat16 to float32, which is not the original input dtype.

eellison replied:
That's what `triton_compute_type` does. not `triton_type`.

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -849,7 +850,11 @@ def expm1(x):
 
     @staticmethod
     def sqrt(x):
```
Since the dtypes are not correct currently - we should use them in codegen yet

arui-meta replied:
I see. The only reason I modified `sqrt` codegen is to have an unittest. I can remove it if it seems risky.

eellison replied:
Sorry, meant to type *should not. Let's still check `codegen_upcast_to_fp32` here just to be safe.

eellison reviewed (commented):
Needs to fix the leaking of inductor into prims dtype promotion before landing, then we can land. 

It's make a bit nervous to have dtype promotion that is not totally correct in main. Can we guard any uses of the dtype in codegen with the config option? @blaine-rister commented on that as well.

Things that needs to be implemented still for correctness:

- using `is_scalar` 
- following `masked` 
- making sure that the triton types and pytorch types align - concretely, we would need to add casts whenever there is a divergence in codegen (such as with triton upcasting div/mod).

eellison commented on torch/_prims_common/__init__.py:
```diff
@@ -1495,7 +1495,7 @@ def _find_highest_dtype_filtered(
         one_plus_dim_tensor_dtype = None
         for x in args:
-            if isinstance(x, TensorLike) and filter(x.dtype):
+            if isinstance(x, (TensorLike)) and x.dtype is not None and filter(x.dtype):
```
Shouldn't need this change

eellison commented on torch/_inductor/codegen/common.py:
```diff
@@ -1747,13 +1754,216 @@ def generate(
+        else:
+            assert hasattr(inp, "dtype")
+
+            # construct a tmp tensor to use dtype promotion util function
+            return torch.empty(0, dtype=inp.dtype)  # torch.empty(0) has ndim == 1
```
`torch.empty(0, dtype=inp.dtype)` actually has different type promotion rules than a tensor because it acts as a scalar. Can you do `torch.empty([1], dtype=inp.dtype)` instead ? Read the docs on type promotion here - https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc

` # torch.empty(0) has ndim == 1` delete this comment  ?

eellison commented on torch/_inductor/codegen/common.py:
```diff
@@ -1723,7 +1730,7 @@ def generate(
         var = self.cache.get(cache_key, None)
         if not var:
-            var = self.newvar(bounds)
+            var = self.newvar(dtype, bounds)
```
Just commenting on this now but.. It's better practice to add a new arg at the end of arguments, typically...

eellison reviewed (approved):
please remove change to `_prims_common` and fix constructing 0-dim tensor for the type promotion

blaine-rister reviewed (approved):
LGTM! Thanks for the new feature.

arui-meta committed:
Add dtype attribute to CSEVariable in Inductor codegen (#136778)

Summary:


- This diff introduces `dtype` attribute to `CSEVariable` and a dtype propagation helper function to infer dtype from input to output for each op.

- In the future, we could use this to do dtype-aware codegen

Test Plan: CI

Reviewed By: blaine-rister

Differential Revision: D61815079

wdvr commented:
@pytorchmergebot merge -f "macos runner issues - pushing this one through"
